### DISCUSSION TITLE
Woo Blocks and Interactivity API: next steps

### DISCUSSION THREAD

**Discussion by gigitux** [woocommerce/woocommerce#44965] [WooCommerce, Blocks, and Site Editing]
Now that we've successfully implemented the Interactivity API for the Product Button, the next step is determining the blocks that would benefit from migrating to the Interactivity API. I've identified two blocks, namely the Mini Cart Block and the Filter blocks, which could yield the following advantages through migration:

- Improve the UX for the final user.
- Reduce the complexity of the codebase.

### Mini Cart Block

The current implementation of the block involves rendering the Mini Cart block on the PHP side ([source code](https://github.com/woocommerce/woocommerce-blocks/blob/045e7b96d0773fb666f1e44447a1ad73dbe43c51/src/BlockTypes/MiniCart.php/#L374-L376)). However, the counter of products and the total price are updated by fetching data from the Store API and manipulating the DOM using vanilla JS ([source code](https://github.com/woocommerce/woocommerce-blocks/blob/045e7b96d0773fb666f1e44447a1ad73dbe43c51/assets/js/blocks/mini-cart/utils/data.ts)). In order to maintain real-time updates of the product count and total price, we replace the PHP markup with a React Block when the user interacts with the Card or hovers over the Mini Cart icon ([source code](https://github.com/woocommerce/woocommerce-blocks/blob/045e7b96d0773fb666f1e44447a1ad73dbe43c51/assets/js/blocks/mini-cart/component-frontend.tsx/#L12-L77)).

Thanks to Interactivity API, we can improve this! Since the Interactivity Store is in sync with the Cart Redux Store, we can utilize selectors to keep the Mini Cart updated, similar to how we handle the Product Button ([source code](https://github.com/woocommerce/woocommerce-blocks/blob/92510d20e47dcfc2a46a7037daf522bfe9535b60/assets/js/atomic/blocks/product-elements/button/frontend.tsx/#L31-L63)). This means that we no longer need to replace the PHP markup with a React Block. Instead, we can leverage the Interactivity API to directly update the Mini Cart using selectors.

By utilizing the Interactivity API and selectors, we eliminate the need for manual DOM manipulation and achieve seamless updates of the Mini Cart. The Interactivity Store, which remains synchronized with the Cart Redux Store, allows us to retrieve the necessary information and keep the Mini Cart up to date without additional JavaScript code.

This improvement simplifies the codebase and ensures a more efficient and streamlined process for updating the Mini Cart. With the Interactivity API handling the updates, we reduce complexity and enhance the overall performance of the block.

TLDR:
- we can reduce the amount of JavaScript that needs to be downloaded on the first interaction with the card and when hovering over it.
- Interactivity API allows us to reduce the complexity of the codebase.
- I estimate that the migration can be completed within one week of work. 

### Filter blocks
Currently, the Filter blocks are implemented as client-side blocks. This approach introduces additional loading time as the JavaScript needs to be executed, and the blocks have to fetch data via the Store API before becoming visible. Moreover, whenever a filter is applied, a complete refresh of the page is required since the Products block (and the new Product Collection block) is a server-side block.

https://github.com/woocommerce/woocommerce-blocks/assets/4463174/3ba848da-7d10-4bb9-abcb-3ce30b3a9a8d

In the short term, we can implement a solution to prevent the need for a complete page refresh. By utilizing the Interactivity API, we can trigger client-side navigation while refreshing only the content of the Products block. This approach allows for a smoother user experience by eliminating the disruptive page refresh.

In the long term, it would be beneficial to migrate the filter blocks to be server-side blocks. This would enable us to update the counters using the selector technique, similar to how we handle the Product Button and I propose for the Mini Cart Block. Doing so can avoid the initial loading time. The combination of client-side navigation and replacing the content of a specific portion of the DOM is still experimental, requiring coordination with the team working on the Interactivity API. Additionally, migrating the filter blocks to server-side blocks presents challenges due to their complexity and size.

TLDR:
- Thanks to Interactivity API, we can avoid the complete refresh of the page.
- In the future, it would be cool to migrate the filter blocks to be server side to avoid the initial loading time.
- The API that triggers the client-side navigation and refreshes a specific portion of DOM is still WIP.
- The filter blocks are very complex, and on the first glue, it is hard doing an estimation.

**Comment by ResponsiveFlow** [comment]
As suggested by @gigitux in woocommerce/woocommerce#44961 I'll add a comment here to explain why the interactivity API should be added to the single product template.

Almost all modern ecommerce stores do not require a page reload when adding a product to the cart.
This is especially important because it reduces the loading time and, most importantly, it enables the Auto-open feature of the mini cart drawer, which I think is a necessary feature for any ecommerce store I build.

It is supported by almost any WordPress classic themes (woostify, kadence, astrawp, etc) through AJAX, and by any themes on shopify.

Since I've made the switch from Classic themes to block themes, in order to have the add to cart + mini cart drawer that opens automatically (which was supported by the classic theme using AJAX) I have rely on a [third party plugin](https://wordpress.org/plugins/woo-ajax-add-to-cart/), which I would rather avoid in favor of an official implementation by WooCommerce itself.
I think this is a basic feature that is available on almost any ecommerce store.

| WITHOUT AJAX | WITH AJAX |
|--------|--------|
| ![ezgif com-optimize-6](https://github.com/woocommerce/woocommerce-blocks/assets/47855768/c888e05d-e9b8-4e9a-8c00-8d259c73a619) | ![ezgif com-optimize-5](https://github.com/woocommerce/woocommerce-blocks/assets/47855768/e061a0ce-4444-461a-8fb0-1d21d3ed2312)
 


This is a [discussion on Slack](https://woocommercecommunity.slack.com/archives/C03MCL9RTAR/p1685458956739979) requesting the exact same feature, but I'm sure there are many more.

I'd love to get a feedback from a designer, maybe @jarekmorawski
